provider/aws/aws.go

```diff
@@ -134,6 +158,69 @@ type Route53API interface {
 	ListTagsForResourceWithContext(ctx context.Context, input *route53.ListTagsForResourceInput, opts ...request.Option) (*route53.ListTagsForResourceOutput, error)
 }
 
+type RateLimitedRoute53Api struct {
+	rateLimiter ratelimit.Limiter
+	client Route53API
+}
+
+func NewRateLimitedRoute53Api(route53api Route53API, rateLimit int) *RateLimitedRoute53Api {
+	return &RateLimitedRoute53Api{
+		rateLimiter: ratelimit.New(rateLimit),
+		client: route53api,
+	}
+}
+
+func (r *RateLimitedRoute53Api) ListResourceRecordSetsPagesWithContext(ctx context.Context, input *route53.ListResourceRecordSetsInput, fn func(resp *route53.ListResourceRecordSetsOutput, lastPage bool) (shouldContinue bool), opts ...request.Option) error {
+	r.rateLimiter.Take()
+	route53RequestsTotal.Inc()
+	err := r.client.ListResourceRecordSetsPagesWithContext(ctx, input, fn, opts...)
+	if err != nil {
+		route53ErrorsTotal.Inc()
+	}
+	return err
+}
+
+func (r *RateLimitedRoute53Api) ChangeResourceRecordSetsWithContext(ctx context.Context, input *route53.ChangeResourceRecordSetsInput, opts ...request.Option) (*route53.ChangeResourceRecordSetsOutput, error) {
+	r.rateLimiter.Take()
+	route53RequestsTotal.Inc()
+	res, err := r.client.ChangeResourceRecordSetsWithContext(ctx, input, opts...)
+	if err != nil {
+		route53ErrorsTotal.Inc()
+	}
+	return res, err
+}
+
+func (r *RateLimitedRoute53Api) CreateHostedZoneWithContext(ctx context.Context, input *route53.CreateHostedZoneInput, opts ...request.Option) (*route53.CreateHostedZoneOutput, error) {
+	r.rateLimiter.Take()
+	route53RequestsTotal.Inc()
+	res, err := r.client.CreateHostedZoneWithContext(ctx, input, opts...)
+	if err != nil {
+		route53ErrorsTotal.Inc()
+	}
+	return res, err
+}
+
+func (r *RateLimitedRoute53Api) ListHostedZonesPagesWithContext(ctx context.Context, input *route53.ListHostedZonesInput, fn func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool), opts ...request.Option) error {
+	r.rateLimiter.Take()
+	route53RequestsTotal.Inc()
+	err := r.client.ListHostedZonesPagesWithContext(ctx, input, fn, opts...)
+	if err != nil {
+		route53ErrorsTotal.Inc()
+	}
+	return err
+}
+
+func (r *RateLimitedRoute53Api) ListTagsForResourceWithContext(ctx context.Context, input *route53.ListTagsForResourceInput, opts ...request.Option) (*route53.ListTagsForResourceOutput, error) {
```

Comment:
If this was jabascript we could have a function that would just wrap all methods calls on the client in a single place, and help us avoid all this copypaste. I assume go doesn't make things like that possible.

Reply:
Eh probably not really, or not without jumping through a bunch of hoops, so whatever. Just an observation.